In [1]:
# import necessary modules
import pandas as pd
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt


from IPython.display import display

### Load the data & preview the data

In [2]:
data = pd.read_csv('../data/raw/Copy of Ask A Manager Salary Survey 2021 (Responses) - Form Responses 1.csv')

In [3]:
print(data.info())
data.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28061 entries, 0 to 28060
Data columns (total 18 columns):
 #   Column                                                                                                                                                                                                                                Non-Null Count  Dtype  
---  ------                                                                                                                                                                                                                                --------------  -----  
 0   Timestamp                                                                                                                                                                                                                             28061 non-null  object 
 1   How old are you?                                                                                             

Timestamp How old are you?  What industry do you work in?  \
0  4/27/2021 11:02:10            25-34   Education (Higher Education)   
1  4/27/2021 11:02:22            25-34              Computing or Tech   
2  4/27/2021 11:02:38            25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41            25-34                     Nonprofits   
4  4/27/2021 11:02:42            25-34  Accounting, Banking & Finance   

                                  Job title  \
0        Research and Instruction Librarian   
1  Change & Internal Communications Manager   
2                      Marketing Specialist   
3                           Program Manager   
4                        Accounting Manager   

  If your job title needs additional context, please clarify here:  \
0                                                NaN                 
1                                                NaN                 
2                                                NaN                 
3                                                NaN                 
4                                                NaN                 

  What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)  \
0                                             55,000                                                                                                                                                                                     
1                                             54,600                                                                                                                                                                                     
2                                             34,000                                                                                                                                                                                     
3                                             62,000                                                                                                                                                                                     
4                                             60,000                                                                                                                                                                                     

   How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.  \
0                                                0.0                                                                                                                                                
1                                             4000.0                                                                                                                                                
2                                                NaN                                                                                                                                                
3                                             3000.0                                                                                                                                                
4                                             7000.0                                                                                                                                                

  Please indicate the currency  \
0                          USD   
1                          GBP   
2                          USD   
3                          USD   
4                          USD   

  If "Other," please indicate the currency here:   \
0                      

### Cleaning data:
- standardize column names
- Remove unnecessary columns
- convert columns to the appropriate data type
- handle missing values

In [4]:
# Make a copy of the original data
df = data.copy()

In [5]:
df.columns

Index(['Timestamp', 'How old are you?', 'What industry do you work in?',
       'Job title',
       'If your job title needs additional context, please clarify here:',
       'What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)',
       'How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.',
       'Please indicate the currency',
       'If "Other," please indicate the currency here: ',
       'If your income needs additional context, please provide it here:',
       'What country do you work in?',
       'If you're in the U.S., what state do you work in?',
       'What city do you work in?',
       'How many years of professional work experience do you have overall?',
       

**Standardize column names**

In [6]:
column_names = {
    'Timestamp': 'timestamp',
    'How old are you?': 'age_range',
    'What industry do you work in?': 'industry',
    'Job title': 'job_title',
    'If your job title needs additional context, please clarify here:': 'job_title_context',
    "What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)":'salary(annualized)',
    'How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.': 'additional_compensation',
    'Please indicate the currency':'currency',
    'If "Other," please indicate the currency here: ': 'other_currency',
    'If your income needs additional context, please provide it here:': 'income_context',
    'What country do you work in?': 'country',
    "If you're in the U.S., what state do you work in?":'state(US)',
    'What city do you work in?': 'city(US)',
    'How many years of professional work experience do you have overall?': 'years_of_experience',
    'How many years of professional work experience do you have in your field?': 'years_of_experience_in_field',
    'What is your highest level of education completed?': 'level_of_education',
    'What is your gender?': 'gender',
    'What is your race? (Choose all that apply.)': 'race'
}
df.columns = df.columns.map(column_names)

In [7]:
df.head()

timestamp age_range                       industry  \
0  4/27/2021 11:02:10     25-34   Education (Higher Education)   
1  4/27/2021 11:02:22     25-34              Computing or Tech   
2  4/27/2021 11:02:38     25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41     25-34                     Nonprofits   
4  4/27/2021 11:02:42     25-34  Accounting, Banking & Finance   

                                  job_title job_title_context  \
0        Research and Instruction Librarian               NaN   
1  Change & Internal Communications Manager               NaN   
2                      Marketing Specialist               NaN   
3                           Program Manager               NaN   
4                        Accounting Manager               NaN   

  salary(annualized)  additional_compensation currency other_currency  \
0             55,000                      0.0      USD            NaN   
1             54,600                   4000.0      GBP            NaN   
2             34,000                      NaN      USD            NaN   
3             62,000                   3000.0      USD            NaN   
4             60,000                   7000.0      USD            NaN   

  income_context         country       state(US)     city(US)  \
0            NaN   United States   Massachusetts       Boston   
1            NaN  United Kingdom             NaN    Cambridge   
2            NaN              US       Tennessee  Chattanooga   
3            NaN             USA       Wisconsin    Milwaukee   
4            NaN              US  South Carolina   Greenville   

  years_of_experience years_of_experience_in_field level_of_education  \
0           5-7 years                    5-7 years    Master's degree   
1        8 - 10 years                    5-7 years     College degree   
2         2 - 4 years                  2 - 4 years     College degree   
3        8 - 10 years                    5-7 years     College degree   
4        8 - 10 years                    5-7 years     College degree   

       gender   race  
0       Woman  White  
1  Non-binary  White  
2       Woman  White  
3       Woman  White  
4       Woman  White

**Dropping Unnecessary Columns Or Columns with High Missing Values**

In [8]:
# drop columns with high missing values
(df.isnull().mean() *100).sort_values(ascending=False)

other_currency                  99.269449
income_context                  89.159331
job_title_context               74.120666
additional_compensation         25.996935
state(US)                       17.896725
level_of_education               0.791134
race                             0.630769
gender                           0.609387
city(US)                         0.292221
industry                         0.263711
job_title                        0.003564
timestamp                        0.000000
age_range                        0.000000
salary(annualized)               0.000000
currency                         0.000000
country                          0.000000
years_of_experience              0.000000
years_of_experience_in_field     0.000000
dtype: float64

'income_context', 'other_currency', 'job_title_context' all have more than 70% missing values, so we can drop them. But before doing so, let's check if there are any meaningful information in these columns.

In [9]:
print('Number of Non-null values in the Other_currency Column: ',len(df[~df['other_currency'].isnull()]))
print('Number of unique values in the Other_currency Column: ',df[~df['other_currency'].isnull()]['other_currency'].nunique())
df[~df['other_currency'].isnull()].iloc[:,:10].sample(20)

Number of Non-null values in the Other_currency Column:  205
Number of unique values in the Other_currency Column:  121


timestamp age_range                              industry  \
11327    4/28/2021 0:25:09     35-44                     Computing or Tech   
9955    4/27/2021 19:29:40     18-24         Education (Primary/Secondary)   
26727   8/26/2021 15:51:01     35-44          Education (Higher Education)   
18947    4/29/2021 5:35:26     25-34                     Computing or Tech   
27766    4/14/2022 7:16:45     25-34          Engineering or Manufacturing   
14072   4/28/2021 14:00:07     35-44                                   Oil   
11332    4/28/2021 0:25:42     25-34           Marketing, Advertising & PR   
16627   4/28/2021 18:17:19     35-44                     Computing or Tech   
3124    4/27/2021 11:59:26     25-34                                 Sales   
1840    4/27/2021 11:33:16     35-44        Intergovernmental organization   
27897  11/10/2022 11:49:07     25-34         Accounting, Banking & Finance   
27216  10/26/2021 13:21:45     25-34         Accounting, Banking & Finance   
27504  11/16/2021 12:01:59     25-34                Business or Consulting   
10978   4/27/2021 22:56:12     25-34                Business or Consulting   
1924    4/27/2021 11:34:55     35-44                                 Sales   
17547   4/28/2021 20:46:27     18-24          Education (Higher Education)   
9284    4/27/2021 17:36:26     45-54                Business or Consulting   
18662    4/29/2021 1:37:56     18-24                     Computing or Tech   
21545   4/30/2021 12:10:47     45-54  Government and Public Administration   
11719    4/28/2021 4:09:53     25-34                     Computing or Tech   

                           job_title  \
11327         Helpdesk Professional    
9955                         Teacher   
26727                   Receptionist   
18947                  game producer   
27766      Graduate Process Engineer   
14072                      Purchaser   
11332      Digital Strategic Planner   
16627        Director of Engineering   
3124               Territory Manager   
1840                      Consultant   
27897        Manager, Risk Reporting   
27216  Planning and forecasting lead   
27504            Executive Assistant   
10978                   Lead Analyst   
1924                             CEO   
17547      Quality Assurance Auditor   
9284   Personal Executive Assistant    
18662      Site reliability engineer   
21545                          Clerk   
11719               Technical Writer   

                                       job_title_context salary(annualized)  \
11327                                                NaN            525,000   
9955   I teach at a training center (extracurricular ...            120,000   
26727                                       Receptionist              40000   
18947            a gamedev equivalent of project manager             60,000   
27766                                                NaN             115000   
14072                                                NaN            689,000   
11332  Position within digital / social media marketi...            720,000   
16627                                                NaN            580,000   
3124              Software sales for automotive industry             33,000   
1840                                                 NaN             60,000   
27897                                                NaN             110000   
27216                                                NaN             180000   
27504                                                NaN              37200   
10978                                                NaN          2,100,000   
1924                                                 NaN            800,000   
17547                                                NaN              2,400   
9284                                                 NaN                 55   
18662                                                NaN            163,200   
21545                               

There are 121 out of 205 unique values in the 'other_currency' column, with makes it a lot of unique values to deal with and the benefit of dealing with each one is minimal. So the current approach at the moment would be to first drop all the rows in the dataframe that have values in the other_currency column because they can skew the analysis of the data then drop the column itself after.

The other columns with high missing values (income_context, job_title_context) can be dropped without losing any relevant information.

In [10]:
# Remove rows where 'other_currency' column has non-null values.
df = df[df['other_currency'].isnull()]
df.shape

(27856, 18)

In [11]:
# Drop unnecessary columns.
df = df.drop(columns=['income_context', 'other_currency', 'job_title_context'],axis=1)

In [12]:
df.head()

timestamp age_range                       industry  \
0  4/27/2021 11:02:10     25-34   Education (Higher Education)   
1  4/27/2021 11:02:22     25-34              Computing or Tech   
2  4/27/2021 11:02:38     25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41     25-34                     Nonprofits   
4  4/27/2021 11:02:42     25-34  Accounting, Banking & Finance   

                                  job_title salary(annualized)  \
0        Research and Instruction Librarian             55,000   
1  Change & Internal Communications Manager             54,600   
2                      Marketing Specialist             34,000   
3                           Program Manager             62,000   
4                        Accounting Manager             60,000   

   additional_compensation currency         country       state(US)  \
0                      0.0      USD   United States   Massachusetts   
1                   4000.0      GBP  United Kingdom             NaN   
2                      NaN      USD              US       Tennessee   
3                   3000.0      USD             USA       Wisconsin   
4                   7000.0      USD              US  South Carolina   

      city(US) years_of_experience years_of_experience_in_field  \
0       Boston           5-7 years                    5-7 years   
1    Cambridge        8 - 10 years                    5-7 years   
2  Chattanooga         2 - 4 years                  2 - 4 years   
3    Milwaukee        8 - 10 years                    5-7 years   
4   Greenville        8 - 10 years                    5-7 years   

  level_of_education      gender   race  
0    Master's degree       Woman  White  
1     College degree  Non-binary  White  
2     College degree       Woman  White  
3     College degree       Woman  White  
4     College degree       Woman  White

**Cleaning other columns with missing values**

In [13]:
(df.isnull().mean() * 100).sort_values(ascending=False)

additional_compensation         25.969271
state(US)                       17.428920
level_of_education               0.771827
race                             0.621051
gender                           0.610281
city(US)                         0.294371
industry                         0.258472
job_title                        0.003590
salary(annualized)               0.000000
timestamp                        0.000000
age_range                        0.000000
currency                         0.000000
country                          0.000000
years_of_experience              0.000000
years_of_experience_in_field     0.000000
dtype: float64

- Additional Compensation: 25% missing values, filling with 0

In [14]:
# fill missing values in 'additional_compensation' column with 0
df['additional_compensation'] = df['additional_compensation'].fillna(0)